### ISSUE TITLE
Refresh authentication token flow

### ISSUE THREAD

**Issue by wouter1975** [workfloworchestrator/orchestrator-ui-library#1423]
**What**
Get the refresh token flow working. 

**Why**
Currently when a token expires a user has to login again to get a new one. For surf that is less of a problem since we have a long expiration time. This issue was encountered by Mohammad from GEANT where they have a expiration time of 1 hour. 

**Acceptance**
Get the refresh token flow working. 
An option might be that we try to refresh the token after we get an unauthenticated error, another option would be to check the validity of the token every time it's used and try to refresh every tim.

NOTES:
- We do an explicit logout() whenever we get a 403 not-authenticated error from the backend. This might interfere with the refresh token flow.
- Mohammad has made an implementation for the refresh token flow. See his code in the comment. He's running into unpredictable behaviour.

Mohammads description of the situation

"Our OIDC IDP issues tokens with a 1-hour expiration. To handle this, we've implemented a refresh token mechanism that generates a new access token, with the refresh token remaining valid for 24 hours. Despite this logic, after staying on the GUI for 1 hour, users are still redirected to the login page, as though the session is not being updated correctly. I understand that Next.js uses both server-side and client-side components, but it appears that somewhere in core-lib code, we are not retrieving the latest session data and are relying on an outdated session instead."

**Comment by DutchBen** [comment]
`import {
  WfoSession,
  WfoUserProfile,
  getEnvironmentVariables,
} from '@orchestrator-ui/orchestrator-ui-components';
import NextAuth, { AuthOptions } from 'next-auth';
import { JWT } from 'next-auth/jwt';
import { OAuthConfig } from 'next-auth/providers';

const {
  OAUTH2_ACTIVE,
  OAUTH2_CLIENT_ID,
  OAUTH2_CLIENT_SECRET,
  NEXTAUTH_PROVIDER_ID,
  NEXTAUTH_PROVIDER_NAME,
  NEXTAUTH_AUTHORIZATION_SCOPE_OVERRIDE,
  OIDC_CONF_FULL_WELL_KNOWN_URL,
  OIDC_TOKEN_ENDPOINT,
} = getEnvironmentVariables([
  'OAUTH2_ACTIVE',
  'OAUTH2_CLIENT_ID',
  'OAUTH2_CLIENT_SECRET',
  'NEXTAUTH_PROVIDER_ID',
  'NEXTAUTH_PROVIDER_NAME',
  'NEXTAUTH_AUTHORIZATION_SCOPE_OVERRIDE',
  'OIDC_CONF_FULL_WELL_KNOWN_URL',
  'OIDC_TOKEN_ENDPOINT',
]);

const isOauth2Enabled = OAUTH2_ACTIVE?.toLowerCase() != 'false';

const token_endpoint_auth_method = OAUTH2_CLIENT_SECRET
  ? 'client_secret_basic'
  : 'none';

async function refreshAccessToken(token: JWT): Promise<JWT> {
  try {
    const raw = JSON.stringify({
      client_id: OAUTH2_CLIENT_ID,
      grant_type: 'refresh_token',
      refresh_token: token.refreshToken as string,
    });

    const response = await fetch(OIDC_TOKEN_ENDPOINT, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: raw,
      redirect: 'follow',
    });

    if (response.ok) {
      const data: {
        access_token: string;
        expires_in: number;
        refresh_token: string;
      } = await response.json();

      return {
        ...token,
        accessToken: data.access_token,
        accessTokenExpires:
          Math.floor(Date.now() / 1000) + (data.expires_in - 300), // Store expiry as a Unix timestamp (in seconds)
        refreshToken: data.refresh_token,
      };
    } else {
      return {
        ...token,
        error: `${response.statusText}`,
      };
    }
  } catch (error) {
    return {
      ...token,
      error: `${error}`,
    };
  }
}

const wfoProvider: OAuthConfig<WfoUserProfile> = {
  id: NEXTAUTH_PROVIDER_ID,
  name: NEXTAUTH_PROVIDER_NAME,
  type: 'oauth',
  clientId: OAUTH2_CLIENT_ID,
  wellKnown: OIDC_CONF_FULL_WELL_KNOWN_URL,
  authorization: {
    params: {
      scope: NEXTAUTH_AUTHORIZATION_SCOPE_OVERRIDE,
      prompt: 'consent',
    },
  },
  idToken: true,
  checks: ['pkce', 'state'],
  userinfo: {
    request: async (context) => {
      const { client, tokens } = context;

      if (!context.provider.wellKnown || !tokens.access_token) {
        return {};
      }
      return await client.userinfo(tokens.access_token);
    },
  },
  profile(profile) {
    return {
      id: profile.sub,
      name: profile.name ?? profile.preferred_username,
      email: profile.email,
    };
  },
  client: {
    token_endpoint_auth_method,
    response_types: ['code'],
  },
};

export const authOptions: AuthOptions = {
  providers: isOauth2Enabled ? [wfoProvider] : [],
  callbacks: {
    async jwt({ token, account, profile }): Promise<JWT> {
      // The "account" is only available right after signing in -- adding useful data to the token
      if (account) {
        return {
          ...token,
          accessToken: account.access_token,
          refreshToken: account.refresh_token,
          accessTokenExpires: account.expires_at as number,
          profile: profile as WfoUserProfile,
        };
      }

      const now = Math.floor(Date.now() / 1000);
      if (
        typeof token.accessTokenExpires === 'number' &&
        now < token.accessTokenExpires
      ) {
        return token;
      }

      return await refreshAccessToken(token);
    },
    async session({ session, token }: { session: WfoSession; token: JWT }) {
      // Assign data to the session to be available in the client through the useSession hook
      session.profile = token.profile as WfoUserProfile | undefined;
      session.accessToken = token.accessToken ? String(token.accessToken) : '';

      return session;
    },
  },
};

export default NextAuth(authOptions);`

**Comment by torkashvand** [comment]
Hi 
I just pasted the message from slack here as well. just in case people are not there.

Last week, I [reported](https://github.com/workfloworchestrator/orchestrator-ui-library/issues/1423) a problem with the refresh token.
Yesterday, I was fixing an issue with the WebSocket and noticed that the GUI (next-auth) sends a big cookie header. I thought maybe NGINX wasn't sending this to the backend.
The issue happened when the WebSocket was working, then after almost 2 weeks suddenly it stopped!!
After checking, I found out that the problem with WebSocket was because of the big cookie header. I changed the NGINX config and removed the cookie from the headers, which wasn’t needed for the Python backend (but it’s still needed for the Node.js backend).
After this, the WebSocket started working again, and surprisingly, the refresh token problem was also fixed. I’m not sure why, but the thing that matters is that it works!
I just wanted to share this with you since I reliezed it is an in-progress task at your side and it may help!
this is my nginx conf for those are interested:
```nginx
http {

    proxy_buffers 4 64k;
    proxy_buffer_size 64k;
    large_client_header_buffers 8 256k;
    client_header_buffer_size 32k;
.
.
.
location /api/ws/events {
   proxy_pass  {{ gso_api_service_url }};
   proxy_set_header Upgrade $http_upgrade;
   proxy_set_header Connection "Upgrade";
   proxy_set_header Sec-WebSocket-Protocol $http_sec_websocket_protocol;
   proxy_set_header Sec-WebSocket-Version $http_sec_websocket_version;
   proxy_set_header Sec-WebSocket-Key $http_sec_websocket_key;
   proxy_send_timeout 3600s;
   proxy_read_timeout 3600s;
   proxy_connect_timeout 10s;
   proxy_buffering off;
   proxy_buffer_size 16k;
   proxy_buffers 4 16k;
   proxy_set_header Cookie "";
}

```
